File: sdk/java/src/main/java/com/anyvali/schemas/AnySchema.java
package com.anyvali.schemas;

import com.anyvali.Schema;
import com.anyvali.ValidationContext;

import java.util.LinkedHashMap;
import java.util.Map;

/**
 * Schema that accepts any value.
 */
public class AnySchema extends Schema<Object> {

    public AnySchema() {}

    private AnySchema(AnySchema other) {
        this.coercion = other.coercion;
        this.defaultValue = other.defaultValue;
        this.hasDefault = other.hasDefault;
    }

    @Override
    protected boolean acceptsNull() {
        return true;
    }

    @Override
    protected Object validate(Object input, ValidationContext ctx) {
        return input;
    }

    @Override
    protected Map<String, Object> toNode() {
        return addCommonNodeFields(new LinkedHashMap<>(Map.of("kind", "any")));
    }

    @Override
    protected Schema<Object> copy() {
        return new AnySchema(this);
    }
}

File: sdk/kotlin/src/main/kotlin/com/anyvali/schemas/AnySchema.kt
package com.anyvali.schemas

import com.anyvali.*
import kotlinx.serialization.json.*

class AnySchema : Schema<Any?>() {
    override val kind: String = "any"

    fun describe(description: String, opts: DescribeOptions? = null): AnySchema {
        applyDescribe(description, opts)
        return this
    }

    fun metadata(meta: Map<String, Any?>, replace: Boolean = false): AnySchema {
        applyMetadata(meta, replace)
        return this
    }

    override fun validateValue(value: Any?, ctx: ValidationContext): List<ValidationIssue> {
        return emptyList()
    }

    override fun exportNode(): JsonObject = buildJsonObject {
        put("kind", JsonPrimitive("any"))
        addMetadataToNode(this)
    }
}
